### DISCUSSION TITLE
Creating policy "permitted locations"

### DISCUSSION THREAD

**Discussion by darxsky** [eclipse-edc/Connector#3919] [Q&A]
Good morning everyone, 
I would like to add an extension to check if the consumer's location is included in a specific location or in a list of allowed locations. It's possible to do it? Could you kindly tell me how to integrate this functionality into the connector via extension?
Thank you.

**Comment by jimmarino** [comment]
I'm not sure how one can reliably check if a client is in a particular location. The only way I can think of is via third-party attestation in a credential (`ClaimToken`) or a lookup of the `participantId`. Take a look at the `IdentityService` implementations for details.

Authoring extensions is extensively documented, including in the samples. Please review those. The discussion board is intended for specific questions regarding the codebase.

**Reply by darxsky** [reply]
Thank you! I apologize for asking the question here. For this type of questions where should I write?

**Reply by paullatzelsperger** [reply]
Hi @darxsky the point was not ask broad and unscoped questions :)
If your goal is to _write an extension_, there is documentation, samples,... You will not get a step-by-step tutorial here.
Determining - in a reliable and verifiable way - a consumer's location is a topic that is outside of EDC, really. If you are talking about _static_ locations, e.g. Headquarter address, then you could do this with VerifiableCredentials. If you mean _dynamic locations_, e.g. GPS position + geofences, then that's an entirely different story. Technically, both would have to be claims on a `ClaimToken`.

**Reply by darxsky** [reply]
Hi @paullatzelsperger ,sorry for the misunderstanding. I wasn't asking how to add an extension to the connector. I found the documentation very clear on how to do this. Yes, I mean static location. I have already created an extension to check if the connector is located in the region indicated by the policy but the property indicating which country the connector is located in is missing, so I was wondering how to add this property to the connector.
Thank you very much for your support.

**Reply by paullatzelsperger** [reply]
if you want to express this in a policy, you'll have to also create a policy evaluation function (check `ContractExpiryCheckFunction` for an example), plus binding it to scopes etc. [doc](https://github.com/eclipse-edc/docs/blob/main/developer/handbook.md#policies)

**Reply by darxsky** [reply]
I have already added the policy evaluation function. My point was with which parameter/property of the connector should I check the right operand of the policy since the connector has no properties like location in the claim token. To do this, I realized that I should work with the claim token. 
Thank you.

**Reply by paullatzelsperger** [reply]
you can basically pick this at your own discretion. The key aspect is, that this information must be stored in the `Claimtoken`, but it is irrelevant, under which key.

**Reply by darxsky** [reply]
Ok perfect. Thank you very much.

**Reply by darxsky** [reply]
Good morning, sorry if I reopen the discussion. I have a doubt. I am using the DAPS extension which generates its own claim once the token is obtained from the DAPS server containing the token information. I wanted to know if there was a way to add a claim, key-value pair, "region": "EU", to the claim returned by DAPS, without modifying the DAPS server code since the getClaims() method, implemented in the connector, returns an immutable collection.
Also, can a participant's location be attached to the client/participant id? For example, 8C:A7:DG:85:E1:keyid:8C:A7:DG:85:E1:region:EU? Or does such a solution not meet connector standards?
Thank you.

**Reply by jimmarino** [reply]
You can implement a `ParticipantAgentServiceExtension`.

**Reply by darxsky** [reply]
Thank you very much. So, I need to add an extension(in the /extensions folder) to my project that overrides the attributesFor() method of ParticipantAgentServiceExtension and then register the created extension via the register method of the ParticipantAgentService class, right?

**Reply by jimmarino** [reply]
You need to create an extension that is loaded via the Java [ServiceLoader](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/ServiceLoader.html) and registers an implementation of `ParticipantAgentServiceExtension` with the `ParticipantAgentService` (which can be injected on the extension class).

**Reply by darxsky** [reply]
Perfect, thank you.
